Your first shader tutorial series

clayjohn opened the pull request:
Related #2228

I am putting this up as a WIP so that people can begin reviewing it. My plan for the series is to have 5 new docs:
- What are shaders (WIP, content not complete)
- Your first CanvasItem shader (complete, ready for review)
- Your first Spatial shader (not started)
- Your first Particles shader (not started)
- Your first visual shader (not started)

cbscribe commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,61 @@
+.. _doc_what_are_shaders:
+
+What are shaders
```
Question mark?

cbscribe commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,61 @@
+lower-level initialization work for you. However, GLSL (the OpenGL Shading Language, which Godot uses)
+is still unintuitive and restricting, especially for users who are used to GDScript.
+
+But what are they
```
Question mark?

cbscribe commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,61 @@
+But what are they
+-----------------
+
+Shaders are a special kind of program that runs on the GPU. 
```
We should probably define "GPU" here.

cbscribe commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,61 @@
+
+Shaders are a special kind of program that runs on the GPU. 
+
+Objects are drawn using an associated shader. The output of the shader is the pixel colors drawn
```
This first sentence is a bit opaque.  I'd remove it and say "The output of the shader is the colored pixels of the object drawn..."

cbscribe commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,61 @@
+
+Objects are drawn using an associated shader. The output of the shader is the pixel colors drawn
+to the viewport. Shaders do not save their results, nor can they be used to draw outside their
+corresponding mesh. The lack of flexibility allows shaders to be incredibly fast. 
```
"mesh" hasn't been defined yet, and doesn't necessarily apply (2D).

"This lack of flexibility..."

cbscribe commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,61 @@
+- cant draw outside mesh
+- cant access other pixels from current pixel (or vertices)
+- cant store previous iterations
+- update on the fly (they can, but they need to be compiled)
```
Remove "cant" on each of these.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+of what shaders are please see :ref:`What are shaders <doc_what_are_shaders>`.
+
+This tutorial will focus on the practical aspects of writing shader programs by walking
+readers through the process of writing a shader with both vertex and fragment functions.
```
In the first doc you're using first person ("This tutorial series will walk you through writing your first shader."), but in this one you're much more formal and third person ("This tutorial will ... walking readers through the process"). I prefer the former, but either way we should be consistent.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+"Icon.png". For new projects, this is the Godot icon. You should now see the icon in the viewport.
+
+Next, look down in the Inspector, under the CanvasItem section, click beside "Material" and select
+"new ShaderMaterial". This creates a new Material. Click on the sphere that appears. Godot currently
```
"New ShaderMaterial"

This creates a new Material resource.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+
+In order to use a shader it must be attached inside a :ref:`Material <class_material>` 
+which must be attached to an object. To draw multiple objects with the same material, 
+the material must be attached to each object.
```
I think it would be good to mention that Materials are Resources. Perhaps also link to https://docs.godotengine.org/en/latest/getting_started/step_by_step/resources.html

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+the material must be attached to each object.
+
+All objects derived from a :ref:`CanvasItem <class_canvasitem>` have a material property.
+This includes all gui elements, sprites, TileMaps, MeshInstance2Ds etc.
```
GUI, Sprites

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+doesn't know whether you are writing a CanvasItem Shader or a Spatial Shader and it previews the output
+of spatial shaders. So what you are seeing is the output of the default Spatial Shader.
+
+Click beside "Shader" and select "new Shader". Finally, click on the new shader resource and the shader
```
"New Shader"

cbscribe commented on tutorials/shading/your_first_shader/index.rst:
```diff
@@ -0,0 +1,27 @@
+Your first shader
+=================
+
```
I didn't realize we could put text in the index.rst - this changes everything! :)

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+
+.. code-block:: glsl
+  
+  shader_type canvas_item
```
Missing `;`

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+Your first fragment function
+----------------------------
+
+The fragment function runs for every pixel covered by a triangle and determine what color that pixel should be
```
It might be clearer to remove the "covered by a triangle" part. It doesn't make sense to the target reader.

determine -> determines

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+----------------------------
+
+The fragment function runs for every pixel covered by a triangle and determine what color that pixel should be
+based on information passed in from the vertex function.
```
May be better not to mention the vertex function just yet. Keeping it to one concept at a time.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+based on information passed in from the vertex function.
+
+They are restricted to the pixels covered by the triangle, that means you cannot use one to, for example, 
+create an outline around a sprite. 
```
Again, this is confusing, because they're looking at a square Godot icon. There is no "triangle" in sight. Yes, it's less precise, but maybe more clear to say they are restricted to the pixels of the image/texture.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+  void fragment(){
+    COLOR = vec4(0.4, 0.6, 0.9, 1.0);
+  }
+
```
It would be good here to explicitly explain that the `COLOR` variable is the output.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+Now let's make things more complex.
+
+There are many inputs to the fragment function that you can use for calculating ``COLOR``.
+``UV`` is one of them. Uv coordinates are specified in your mesh and they tell the shader
```
It's not a word, so stick to "UV"

Again, it's not yet clear what a "mesh" is, and in this case we could still use "texture".

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+
+In the fragment function you can only read from ``UV``, but you can use it in other functions
+or to assign values to ``COLOR`` directly.
+
```
I think we need a diagram of what UV coordinates represent and that they range from 0 - 1.

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+Interacting with shaders from code
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can change uniforms from code using the function ``set_shader_param`` which is called on the nodes
```
Function names should include the parens: `set_shader_param()`

nodes -> node's

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can change uniforms from code using the function ``set_shader_param`` which is called on the nodes
+material resource. With a sprite node, the following code can be used to set the ``blue`` uniform.
```
sprite -> Sprite

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+    COLOR.b = blue;
+  }
+
+Now you can change the amount of blue in the sprite from the editor. Look back at the Inspector
```
sprite -> Sprite

cbscribe commented on tutorials/shading/your_first_shader/your_first_canvasitem_shader.rst:
```diff
@@ -0,0 +1,212 @@
+Using ``TEXTURE`` built-in
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When you want to adjust a color in a sprite you cannot just adjust the color from the texture
```
sprite -> Sprite

clayjohn commented:
Just added a two part "your first spatial shader" tutorial. It is pretty long. I think at this point it is best to focus on reviewing these tutorials and then get them merged. I will create a new PR for the next two tutorials. This is just getting to be a lot for one PR. 

@cbscribe I think I addressed all your comments on the first couple tutorials! I really appreciate the feedback.

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+
+Shaders are a special kind of program that runs on Graphics Processing Units (GPUs). All computers
+have some sort of GPU, either one integrated into their CPU or discrete (meaning it is a separate 
+hardware component). GPUs are especially useful for rendering because they are optimized for running
```
I'd change it to "most computers", also explicitely say "graphics card" somwhere - e.g. "meaning it is a seperate hardware component, for example the typical graphics card" - or something. 

If you're new to this, graphics card may be a term you're (already) familiar with.

clayjohn replied:
I agree. Thanks!

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+Shaders are a special kind of program that runs on Graphics Processing Units (GPUs). All computers
+have some sort of GPU, either one integrated into their CPU or discrete (meaning it is a separate 
+hardware component). GPUs are especially useful for rendering because they are optimized for running
+thousands of instructions at once.
```
at once = in parallel

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+hardware component). GPUs are especially useful for rendering because they are optimized for running
+thousands of instructions at once.
+
+The output of the shader is the colored pixels of the object drawn to the viewport. Shaders operate
```
>The output of the shader is the colored pixels of the object drawn to the viewport

This is true for many typical graphics shaders, but it doesn't need to be colored (think greyscale, or black and white), and it doesn't need to be pixels - think OpenCL/CUDA calculations etc.

clayjohn replied:
One the one hand, its nice to keep it simple as this is meant to go along with with "your first shader" tutorials. On the other hand it is better to be more correct. I added "typically" then also added a line saying that other outputs are possible.

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+thousands of instructions at once.
+
+The output of the shader is the colored pixels of the object drawn to the viewport. Shaders operate
+with a fixed-functionality pipeline. The standard process is the vertex -> fragment shader pipeline. The vertex
```
1. Its fixed-function pipeline
2. Shaders were the evolution of the fixed-function pipeline. https://gamedevelopment.tutsplus.com/articles/the-end-of-fixed-function-rendering-pipelines-and-how-to-move-on--cms-21469

clayjohn replied:
I was using fixed-functionality to refer to the vertex -> fragment processing pipeline. Which is technically called the shader pipeline. But in a sense it is a pipeline with a fixed functionality, but not the fixed-function pipeline. Clearly this was confusing. I have tried to clear it up and use alternative language throughout.

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+You have no control over how this function is called. So you have to design your shaders
+differently from how you would design programs on the CPU.
+
+A consequence of this fixed-funtionality is that you cannot access the results from a previous 
```
I think this actually confused what *fixed-functionality* is in a graphics context.

A fixed function pipeline is one where there aren't any shaders that the GPU executes; the GPU state is instead manipulated directly via an API, instead of executing a shader.

>The term Fixed Function Pipeline commonly refers to the set of configurable processing state present in older versions of OpenGL that have since been replaced by Shaders of some sort. While the current OpenGL pipeline still retains non-programmable state, that is not typically what people are referring to when they say "fixed function".
https://www.khronos.org/opengl/wiki/Fixed_Function_Pipeline

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+
+A consequence of this fixed-funtionality is that you cannot access the results from a previous 
+run of the shader, you cannot access other pixels from the pixel being drawn, and you cannot 
+write outside of the current pixel being drawn This lack of flexibility is designed to work with 
```
Missing "." before "This"

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+A consequence of this fixed-funtionality is that you cannot access the results from a previous 
+run of the shader, you cannot access other pixels from the pixel being drawn, and you cannot 
+write outside of the current pixel being drawn This lack of flexibility is designed to work with 
+the GPU which allows shaders to be incredibly fast.
```
I'd go with "this enables the GPU to execute the shader for different pixels *in parallel*, as they do not depend on each other". *This* is what greatly speeds this up.

I mean, this is basically what you wrote, just maybe more explicit?

mhilbrunner commented on tutorials/shading/your_first_shader/what_are_shaders.rst:
```diff
@@ -0,0 +1,111 @@
+
+GPUs are able to render graphics much faster than CPUs for a few reasons, but most notably,
+because they are able to run calculations massively in parallel. A CPU typically has 4 or 8 cores
+while a GPU typically has hundreds. That means a GPU can do hundreds of tasks at once. GPU architects
```
Talking about cores: hundreds = thousands. <1k are already low end today, and will be even more so with each passing day. :)

mhilbrunner commented on tutorials/shading/your_first_shader/your_first_spatial_shader.rst:
```diff
@@ -0,0 +1,340 @@
+you have decided to write your own and now you need figure out where to start.
+
+This tutorial will explain how to write a Spatial shader and will cover more topics than the
+CanvasItem tutorial.
```
Link to the tutorial here?

clayjohn commented:
@mhilbrunner I have addressed all your comments! They were very helpful, thank you.

mhilbrunner commented:
Looks good to me. Maybe we should redirect the old *Vertex displacement with shaders* tutorial page to this new one, as it is getting removed?

clayjohn commented:
How would we do that?

mhilbrunner commented:
Ah that was more a note to myself. As it stands, it seems we can't do version/branch specific redirects anyway, so we'll just leave it like that for now. Merged.

Thanks for all the work, this is impressive! :)
